### PR TITLE
Add uimuldw function

### LINKED ISSUE #207: RFC: add support for computing the double word product of two unsigned 32-bit integers
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes to add support for computing a double word product of two unsigned 32-bit integers. Adding support would partially address lack of 64-bit integer support in JavaScript (ignoring BigInt), which is necessary for, e.g., 32-bit integer multiplication, the product of which may exceed the maximum safe integer when evaluated in double-precision floating-point format.

Package: `@stdlib/math/base/special/uimuldw`
Alias: `uimuldw`

The API would follow the convention set forth of providing an optional output argument:

```javascript
/**
* @param {ArrayLikeObject} [out] - output array
* @param {uinteger32} a - first unsigned integer
* @param {uinteger32} b - second unsigned integer
* @returns {ArrayLikeObject} output array
*/
function uimuldw( [out,] a, b ) {
    // Implementation...
}
```

## Related Issues

> Does this feature request have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

-   [reference C implementation via _Hacker's Delight_](http://www.hackersdelight.org/hdcodetxt/muldwu.c.txt) (originally from Knuth and based on [Montgomery multiplication](http://www.hackersdelight.org/MontgomeryMultiplication.pdf))
-   Related SO issues [[1](https://stackoverflow.com/questions/31652875/fastest-way-to-multiply-two-64-bit-ints-to-128-bit-then-to-64-bit), [2](https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication), [3](https://stackoverflow.com/questions/28766755/compute-the-doubleword-product-signed-of-two-words-given-the-lower-word-produc?lq=1)]

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

### PR DESCRIPTION
**Pull request by rei2hu**
Resolves #207  .

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   Implements the uimuldw function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #207 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

**Comment by kgryte** [comment]
@rei2hu This is awesome! Will review later today!

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# uimuldw
+
+> Perform multiplication of two unsigned 32-bit integers and return an array of two unsigned 32-bit integers which represents their 64-bit unsigned integer product.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
> Compute the double word product of two unsigned 32-bit integers.
```

The current description is **good**; however, a bit too literal for a top-level package description. It's use below can be left as is.

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+```javascript
+var uimuldw = require( '@stdlib/math/base/special/uimuldw' );
+```
+
+<!-- eslint-disable no-shortcut-reference-link -->
+
+#### uimuldw( \[out,\] a, b )
+
+> Perform multiplication of two unsigned 32-bit integers and return an array of two unsigned 32-bit integers which represents their 64-bit unsigned integer product.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
> Multiplies two unsigned 32-bit integers and returns an array of two unsigned 32-bit integers which represents the 64-bit unsigned integer product.
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+
+{{alias}}( [out,] a, b )
+    Perform multiplication of two unsigned 32-bit integers and return an array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt]:
```suggestion
    Multiplies two unsigned 32-bit integers and returns an array
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    of two unsigned 32-bit integers which represents their 64-bit unsigned
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt]:
```suggestion
    of two unsigned 32-bit integers which represents the 64-bit unsigned
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    integer product.
+
+    Parameters
+    ----------
+    out: Array (optional)
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt]:
```suggestion
    out: ArrayLikeObject (optional)
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    a: integer
+        Unsigned 32-bit integer.
+
+    b: integer
+        Unsigned 32-bit integer.
+
+    Returns
+    -------
+    out: Array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt]:
```suggestion
    out: ArrayLikeObject
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/examples/index.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+var uimuldw = require( './../lib' );
+
+var i;
+var j;
+var y;
+
+for ( i = 0xFFFFFFF0; i < 0xFFFFFFFF; i++ ) {
+	for ( j = i; j < 0xFFFFFFFF; j++) {
+		y = uimuldw( i, j );
+		console.log( '%d x %d = 0x%s%s', i, j, y[0].toString( 16 ).padStart( 8, 0 ), y[1].toString( 16 ).padStart( 8, 0 ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/examples/index.js]: `padStart` is not supported in older environments. Would you mind updating to use [`@stdlib/string/left-pad`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/string/left-pad)? Thanks!

File: lib/node_modules/@stdlib/math/base/special/uimuldw/lib/index.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+/**
+* Perform multiplication of two unsigned 32-bit integers and return an array of two unsigned 32-bit integers.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/lib/index.js]:
```suggestion
* Perform multiplication of two unsigned 32-bit integers and return an array of two unsigned 32-bit integers which represents the 64-bit unsigned integer product.
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/lib/main.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+// VARIABLES //
+
+// Define a mask for the least significant 16 bits (low word): 65535 => 0x0000ffff => 00000000000000001111111111111111
+var LOW_WORD_MASK = 0x0000ffff>>>0; // asm type annotation
+
+
+// MAIN //
+
+/**
+* Performs multiplication of two unsigned 32-bit integers and returns an array of two unsigned 32-bit integers.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/lib/main.js]:
```suggestion
* Performs multiplication of two unsigned 32-bit integers and returns an array of two unsigned 32-bit integers which represents the 64-bit unsigned integer product.
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/lib/main.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+* @param {ArrayLikeObject} [out] - output array
+* @param {uinteger32} a - integer
+* @param {uinteger32} b - integer
+* @returns {ArrayLikeObject} output array
+*
+* @example
+* var v = uimuldw( 0xAAAAAAAA, 0x55555555 );
+* // returns [ 954437176, 1908874354 ]
+*/
+function uimuldw() { // [out] a, b
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/lib/main.js]:
```suggestion
function uimuldw() {
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/package.json [hunk 11]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/math/base/special/uimuldw",
+  "version": "0.0.0",
+  "description": "Perform multiplication of two unsigned 32-bit integers and get their 64-bit product.",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/package.json]:
```suggestion
  "description": "Compute the double word product of two unsigned 32-bit integers.",
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/package.json [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    "multiplication",
+    "multiply",
+    "mult",
+    "integer",
+    "unsigned",
+    "uint",
+    "uint32",
+    "32-bit",
+    "32bit",
+	"64-bit",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/package.json]:
```suggestion
    "64-bit",
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/package.json [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	"64bit",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/package.json]:
```suggestion
    "64bit",
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+/**
+* Computes the doubleword product of two words, unsigned.
+*
+* From: http://www.hackersdelight.org/hdcodetxt/muldwu.c.txt
+*
+* @param w output array
+* @param u first word
+* @param v second word
+*/
+void muldwu1(unsigned w[], unsigned u, unsigned v) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c]:
```suggestion
void muldwu1( unsigned w[], unsigned u, unsigned v ) {
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+		// write an unnamed array
+		fprintf( f, "[" );
+		write_array_ui32( f, &y[ i * 2 ], 2 );
+		fprintf( f, "]" );
+		if ( i < len - 1 ) {
+			fprintf( f, "," );
+		}
+	}
+	fprintf( f, "]" );
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/test/fixtures/c/runner.c]:
```suggestion
```

**Review by kgryte — CHANGES_REQUESTED** [review]
@rei2hu This is really good work! Thanks for working on this! Only minor nitpicks. Once resolved, should be ready for merge!

**Comment by rei2hu** [comment]
I'll also pr for the signed version once this passes

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+v = uimul( 0x80000000, 0x80000000 ); // 2^31 * 2^31 = 4611686018427388000 => 32-bit integer overflow
+// returns [ 1073741824, 0 ]
+```
+
+</section>
+
+<!-- /.usage -->
+
+<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
```

This is a stray comment which has wandered from its original place. You can just go ahead and remove.

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+<section class="usage">
+
+## Usage
+
+```javascript
+var uimuldw = require( '@stdlib/math/base/special/uimuldw' );
+```
+
+<!-- eslint-disable no-shortcut-reference-link -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+v = uimul( 0x80000000, 0x80000000 ); // 2^31 * 2^31 = 4611686018427388000 => 32-bit integer overflow
+// returns [ 1073741824, 0 ]
+```
+
+</section>
+
+<!-- /.usage -->
+
+<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/README.md [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt [hunk 22]
```diff
@@ -0,0 +1,30 @@
+
+{{alias}}( [out,] a, b )
+    Multiplies two unsigned 32-bit integers and returns an array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/uimuldw/docs/repl.txt]: This description needs to be re-aligned with 80 char width. I believe `of two` can be moved to this line, which may affect subsequent lines.

**Review by kgryte — CHANGES_REQUESTED** [review]
@rei2hu I missed a couple nits in documentation files. Sorry about that. Once these are resolved, I'll merge. :)

**Review by kgryte — APPROVED** [review]
@rei2hu Thanks again for working on this!

**Pull request merged**